Automattic/wp-calypso
# Add Support Search to Customer Home

As per #38860, embeds a "Support Search" section directly onto the Customer Home screen.

This allows users to search support docs directly without having to toggle open the existing floating FAB icon.

Fixes https://github.com/Automattic/wp-calypso/issues/38860

## Screenshots

![Screen Capture on 2020-05-22 at 14-32-19](https://user-images.githubusercontent.com/444434/82673124-456c0f00-9c39-11ea-9d53-301d99f25758.gif)

## Testing instructions

* Checkout this PR.
* Boot Calypso locally. `yarn start`. Wait for a long time...
* Navigate to your Site's Customer Home - `http://calypso.localhost:3000/home/{YOUR_TEST_WEBSITE}`
* See Support search "card" as per design in https://github.com/Automattic/wp-calypso/issues/38860#issuecomment-631125518
* Switch to another language and check all UI strings (not search results as this PR doesn't control those) are translated.
* Proceed to testing search and FAB (see below).

### Search testing

Try out the following scenarios to validate searching behaves as expected. For each scenario you _must_ be able to:

1. Click on the search result to open in modal.
2. Use keyboard down arrow to move from the search input into the results and hit ENTER to open the result in the modal.

The scenarios are:

* Search for empty string - should see contextual results immediately (no loading state).
* Search using a string containing multiple SPACE chars only - should see contextual results immediately (no loading state).
* Search for "Podcast" and see API results.
* Clear search input using "X" icon - should see contextual results immediately (no loading state).
* Clear search input using backspace - should see contextual results immediately (no loading state).

### Validate FAB still works

The Customer Home Support Search reuses many of the same components as the original FAB "Help" floating icon. Therefore it is critical we validate the changes made in this PR have not broken that implementation.

Please repeat the above steps but this time navigate to another section of Calypso such as "Pages". Then click on the "?" FAB icon and check that all the of the above functionality works. The only exception should be that the FAB doesn't not _immediately_ open a modal but shows an interstitial view first within the popover.

### Validate Events Fire

* Open devtools console.
* Paste `localStorage.setItem( 'debug', 'calypso:analytics*' );` and hit `ENTER`.
* Refresh the page.


## Todo

- [x] Use existing `getContextualHelpResults` selector within `requestInlineHelpSearchResults` creator to avoid importing all sorts of helpers. Ideally this means we don't need to pass the current route path into the action creator dispatch call.
- [x] Add comments to explain approach to storing contextual results in state and also the purpose of the state flag which indicates whether results came from the API.
- [ ] Raise Issue in Calypso repo regarding a11y issue with `components/search` whereby input loses focus when it is reset via the "X" icon. 
- [x] Add translation for `No results` within `client/blocks/inline-help/inline-help-search-results.jsx`.
- [x] Hide existing FAB icon on Customer Home.
- [x] Add CSS to fix the height of the box.
- [x] Immediately dispatch contextual results if search query is empty.
- [ ] Don't track searches for empty strings.
- [ ] Ensure we differentiate in tracking between the floating FAB and the Customer Support Search.
- [ ] Consider [using route to determine whether the FAB is shown](https://github.com/Automattic/wp-calypso/blob/24bd44e99c33c4b511eaaedbaf6f654c93c38916/client/layout/index.jsx#L106).
